### DISCUSSION TITLE
CamAPS FX follower

### DISCUSSION THREAD

**Discussion by buxxdev** [NightscoutFoundation/xDrip#2688] [Ideas]
Hi all,
This week a new feature for CamAPS FX was released allowing to download the app on a follower phone and setup as “companion” mode based on an invite from the main user, see here - https://youtu.be/q-dRCGMZA_E

Unfortunately there is no iOS app available so many followers are unable to use this app.  This has coincided with them introducing up to and hour delay in data upload to diasend so there is no method for real-time monitoring of basal, carbs, bolus etc.

My “idea” is to ask if it would be possible to investigate how the app is working and see if xdrip+ could act as a follower as the companion app does and upload into nightscout?

**Comment by Navid200** [comment]
If you can get the app to show notifications, xDrip can already get the readings shown in the notifications.  So, what you are asking for already exists.

**Reply by buxxdev** [reply]
Yes, im already using this.  The ask is for the basal, bolus carb information.

**Reply by Navid200** [reply]
This means the master phone needs to have xDrip installed and collect from CamAPS.  Then, xDrip can share through either xDrip sync or Nightscout.

**Reply by Navid200** [reply]
So, CamAPS developers don't listen to your needs.  Why do you use it?
Is there no alternative?

**Reply by buxxdev** [reply]
We’re using CamAPS after 2 years of using AAPS with my daughter who is now 5 years old.  While there are pros and cons we find CamAPS takes some of the burden away from the regular changes for a child - this is our personal choice.
Only last week there was change in that they will no longer upload real-time data into diasend - no official reason but seems to be because of load reasons in diasend side.
They followed up this week with the release of the companion mode android app which provides all information, but there is no iOS app until the end of the year.  Leaving lots of carers like ourselves who are iOS users without a real-time monitoring solution other than BG which xdrip+ is able to provide us already (thanks 😊).

I added it as an idea as currently I feel at a loss as are many other carers, if there is no appetite though I'm happy to close the thread.

**Reply by Navid200** [reply]
Have you enabled Companion Bluetooth?  Just wondering.  I have no idea if it can read what you are after.

Does CamAPS show insulin and carb info in its notifications?

**Reply by Navid200** [reply]
And there is no need to close a discussion.  All questions, suggestions, requests are welcome.

**Reply by buxxdev** [reply]
Thanks.
The YouTube video explains it, that’s why I posted it.
Basically from the main app you add an email for a follower (a bit like dexcom share).
Then the follower creates an account with this email and log into the app on a separate remote phone.  When you log in and select “companion mode” it knows your account has been invited to follow.  You accept the invite and then the follower phone “companion” app looks the same as the main app but with features disabled (read-only).  I guess you understand now no Bluetooth is involved.

Unfortunately only BG is shown in the notifications so it can’t be taken from there.

**Reply by Navid200** [reply]
I'm afraid we may have a misunderstanding here.

xDrip has a feature called Companion app mode.  
People who use apps that provide blood glucose readings, can also install xDrip and use that feature to get readings from the notifications.
If xDrip gets readings, it can share its readings with a follower.

The companion app mode (in xDrip) can also do back-fill in some cases.  For example, the phone may need to be Android 12 or higher.
I was just wondering if xDrip may be able to use the same mechanism it uses to get old readings, through Bluetooth, to also get insulin or carb values.
But, I was focused on your daughter's phone and you are focused on your phone.  And that is the cause of the misunderstanding.

If xDrip on your daughter's phone can get the insulin and carb info, that xDrip can share it with any other xDrip like the one on your phone.

**Reply by buxxdev** [reply]
Yes I get your point, but in CamAPS all data in in the app. 
i was thinking more along the lines on how the web follower works for libre 3.  On reflection maybe this doesn’t need to be in xdrip, and if it’s possible to act as this CamAPS companion it could be done using a script or nightscout plugin directly.

**Reply by burnedikt** [reply]
Hi, I'm the maintainer of [diasend-nightscout-bridge](https://github.com/burnedikt/diasend-nightscout-bridge) which worked great up until diasend was decomissioned and replaced by glooko. Nowadays, for CamAPS users, I see three potential ways to somehow get the data of CamAPS Fx into nightscout: 

1. Setup a "fake" companion which receives all the data of the CamAPS Fx companion and forwards it to nightscout (not to be confused with xDrip's companion mode, which only reads data from notifications, whereas CamAPS Fx Companion mode mirrors all its data to another device). I have shortly tried this but didn't find a way to "intercept" the data sent between the two devices
2.  Retrieve the data somehow from glooko.com and forward it to nightscout - glooko per se offers an API but the reduced availability of data via their user interface combined with the fact that the data is only uploaded very hour or so, does not make this a great option
3. Intercept the outgoing data on the device via an app like xDrip and forward it to nightscout. Not very promising as it would probably require root access to the Android device and even then may or may not work

**Reply by danbarker** [reply]
Hi @burnedikt, just checking to see if you have had any more thoughts about this since the last comment? We have started using CamAPS and need a way to follow on iOS. I was thinking of making a cloud-hosted follower that would run the official CamAPS follow app  (eg in a virtual android machine?) and read the data out from it, eg with OCR? And then I was thinking of making a follow app but having now seen nightscout, perhaps can just use that! That's about as far as I got, would love any input. Another thought was to try reverse engineering the follow app eg MITM it to find its underlying APIs

**Reply by burnedikt** [reply]
@danbarker For reference, my setup currently looks as follows: 

- Android Phone running CamAPS
- xDrip running as [Companion App](https://xdrip.readthedocs.io/en/latest/install/companion/) (not to be confused with CamAPS Companion) on the same Android Phone, thereby collecting the CGM data directly from CamAPS and sending it to ...
- ... nightscout instance
- multiple iOS devices that simply consume the data from nightscout via either the native nightguard app or the nightscout web app (xDrip for iOS would also be an option and there are a [ton of other iOS clients / "downloaders"](https://nightscout.github.io/nightscout/downloaders/#ios) that can show the data from nightscout out there)

This gets us into the situation of seeing the CGM values in near-realtime. However, all treatments (boli, carbs) are missing. To get those, the only viable option right now is to use [nightscout-connect](https://github.com/nightscout/nightscout-connect#glooko) and configure it to load the treatment data from glooko. The downside here is, that there's a huge delay in the data since the glooko api is not intended for real-time use and CamAPS only sends data to glooko once per hour or so. So for near-real-time monitoring, this is not sufficient and the reason why I don't use that solution right now. 

For near-real-time monitoring, the best solution in my eyes would be to reverse-engineer the API used by CamAPS Companion Mode since that has all the required data in real-time. Running a second virtual or physical android device with CamAPS as you proposed could work, but I don't see that it'll be easy, reliable and cheap to extract the data from this Companion (e.g. via OCR like you proposed). I tried to reverse-engineer the CamAPS API through MITM but was not successful, likely due to certificate pinning within CamAPS. To be fair, I also did not invest a lot of time into it but it's still the most promising approach imho. There's of course also the option to get an additional Android Phone and just use it additionally to the iOS device, just as a CamAPS Companion. 

Personally, I am not investing a lot of time into this anymore as I'm fine just being able to monitor the CGM (although of course treatments would be great to see, it's not a hard requirement, contrary to my initial beliefs). Also, the issue might solve itself anyhow once CamAPS becomes available for iOS in which case you can just use the CamAPS companion mode.

**Reply by danbarker** [reply]
Thanks for your response! 

I managed to get XDrip set up. I didn't initially understand from the thread above that you are able to run XDrip and CamAPS simultaneously on the same phone - although enough people have tried to explain that above, I think this may have been partly because i only saw a snippet of the thread when coming in from Google on mobile. So hopefully your explanation will make it even clearer for anyone else coming in if they make the same mistake I did :)

I'm going to get Nightscout set up too. Just having CGM should be enough for now, I was just very motivated yesterday as we are in the hospital and have switched over, losing our Dexcom Follow ability. And hopefully they do bring out the iOS app soon enough!

**Comment by Navid200** [comment]
@burnedikt
I have 2 questions and one comment. 
1- Is the user carrying only one device for all the three options you have mentioned?
2- Is there anything wrong with using the xDrip companion app mode on the same device that runs CamAPS, and why that is not one of your options?

If the user can carry two devices, the other device can have xDrip on it connected to the medical slot of a G6 and sending the data to Nightscout.

**Reply by burnedikt** [reply]
> 1- Is the user carrying only one device for all the three options you have mentioned?

Yes, the user only carries one device which runs the CamAPS Fx app, which in turn connects to a Dexcom g6 and the insulin pump (Dana I pump or ypsomed pump).

> 2- Is there anything wrong with using the xDrip companion app mode on the same device that runs CamAPS, and why that is not one of your options?
> If the user can carry two devices, the other device can have xDrip on it connected to the medical slot of a G6 and sending the data to Nightscout.

Via xdrip's companion mode it's possible to "scrape" the cgm readings but not the treatments or temp basal adjustments that the CamAPS loop frequently issues. These treatments go unnoticed as they're not published as Android notifications. However, the treatments are quite relevant in a remote monitoring scenario, e.g. for kids. 

Personally, I don't think xdrip can help here more than what it already does (intercepting the cgm values from the notifications) as the other relevant data (treatments) is simply not accessible. It's either sent to glooko.com (every hour or so) and to any CamAPS Companion. Also to clarify, for anyone on an android phone, the CamAPS companion is the logical choice as it shows all the data natively and without any 3rd party solution, so the "problem" mostly affects caretakers who use iOS devices personally. 

The fourth option that I might have forgotten to add previously is to make the caretakers adopt Android instead of iOS ;-)

**Reply by BILLYEDIT** [reply]
Hello, we are using the CamAPS app alongside xdip on our master phone and follow phones.  Its all works very well, on a pixel watch too. :)    Xdrip is fantastic, with its predictions etc, it would be amazing if it _could_ access the additional camAPS info (carbs and bols).  Here's hoping it will happen one day.  Leaving this hello in case something devlelops in the future.  
All the best.
BB

**Comment by beschikbaar-online** [comment]
> Hi all,
> 
> This week a new feature for CamAPS FX was released allowing to download the app on a follower phone and setup as “companion” mode based on an invite from the main user, see here - https://youtu.be/q-dRCGMZA_E
> the youtube video is not
For anyone available It's privatie<
> 
> 
> Unfortunately there is no iOS app available so many followers are unable to use this app.  This has coincided with them introducing up to and hour delay in data upload to diasend so there is no method for real-time monitoring of basal, carbs, bolus etc.
> 
> 
> 
> My “idea” is to ask if it would be possible to investigate how the app is working and see if xdrip+ could act as a follower as the companion app does and upload into nightscout?

**Reply by Navid200** [reply]
These forums are for everyone who need/want to talk about xDrip.  
There is no reason to include a link that requires an invitation and not visible to everyone.

**Comment by bmcnightscout** [comment]
Bit late to this party, but following the release of Camaps for iOS am hoping someone might be able to help :)

We're basically doing what is described above, to get data from our T1D (on Android) to us (on iOS)

- Xdrip set as a companion, reads the notifications and sends the data to Nightscout. This in turn feeds sugarmate (which we have on our iOS devices) and Sugarpixel (which wakes the T1D up!)

We now want to switch the main device to iOS and am trying to figure out a way of getting the data into nightscout, given there is no iOS companion mode and no Xdrip for iOS (without an apple dev account)

So my question is

Could this work if we have the existing Android as a companion to the new iOS Camaps app. That is, the Android acts as a 'companion bridge', and in turn Xdrip reads the notifications from the now secondary companion device and sends them to nightscout as it does now?

Or is there another/better way to do this?

**Reply by beschikbaar-online** [reply]
Download the CamAppsFX app via VPN from Zweden on iOS

**Reply by bmcnightscout** [reply]
That's a great idea and solves my sugarmate problem but not sure how i'd get data into nightscout from iOS?

**Reply by beschikbaar-online** [reply]
You’re right downloading CamAPS FX via the Swedish App Store solves the Sugarmate issue, but unfortunately iOS doesn’t support sending data directly to Nightscout. One possible workaround is to keep your Android device as a “companion bridge” letting xDrip read notifications (if any) from CamAPS on iOS, and forward them to Nightscout like before. It’s not guaranteed to work though, as iOS notification data is limited. Alternatively, you might explore syncing via Apple Health and a custom bridge, but that takes more setup.

**Reply by buxxdev** [reply]
Hi, many users like myself have moved to ios for CamAPS FX and like you suggest we just change the original android phone into a companion mode phone, with xdrip+ running there.  the iOS app is able to send to android app running in companion mode.  currently there is no companion mode available for ios (although irrelevant for this discussion).

I'm not sure how the "sugarmate issue" is solved by downloading CamAPS FX via the Swedish app store?  could someone clarify that?

**Reply by beschikbaar-online** [reply]
Exactly iOS currently doesn’t support Companion Mode, so xDrip+ has to run on an Android device in that role. The idea behind using the Swedish (or Dutch) App Store is that CamAPS FX isn’t available in all countries, so switching App Store regions allows iOS users to download it at all.

Regarding Sugarmate: the app relies on Apple Health data, which CamAPS FX on iOS can write to unlike Android, where CamAPS doesn’t integrate with HealthKit. So by using CamAPS FX on iOS, glucose values are pushed into Apple Health, and from there, Sugarmate can display them.

**Reply by buxxdev** [reply]
Ok this is very interesting, i was not aware that CamAPS FX is writing into Apple Health... is it realtime?

**Reply by beschikbaar-online** [reply]
Yes, CamAPS FX on iOS writes CGM data into Apple Health roughly every 5 minutes. That’s how Sugarmate can access it again as long as Health permissions are enabled. It doesn’t write insulin data though, just glucose.

**Reply by buxxdev** [reply]
Could you explain how to enable the permissions as I don’t see CamAPS in the health app as an app source?

**Reply by Navid200** [reply]
This is the first time I hear that iOS cannot upload to Nightscout.
In fact, I know there is a group called Loop and Learn that is about closed loop systems entirely running on iOS and they use Nightscout.

Please join the group and I am sure they will be able to help you:
https://www.facebook.com/groups/LOOPandLEARN

This repository is about xDrip, which runs only on Android.

**Comment by bmcnightscout** [comment]
Thanks for the replies. So to confirm, with the Android running as a companion, and not the 'main' device, it's still possible to send the notifications from the companion phone to nightscout via xdrip? If so, that solves the problem for me, and the teenager gets an iPhone :) 

I have sugarmate set up to read from nightscout not apple health and we (the parents) use that as our follow/companion app as it comes with the benefit of automated calls overnight for low blood sugar events. Our teen sleeps through all but the sugar pixel alarms (which are also configured via nightscout)

**Reply by buxxdev** [reply]
thats exactly what we're doing currently 👍
